numpy/core/src/multiarray/common.c

```diff
@@ -852,3 +853,102 @@ _may_have_objects(PyArray_Descr *dtype)
     return (PyDataType_HASFIELDS(base) ||
             PyDataType_FLAGCHK(base, NPY_ITEM_HASOBJECT) );
 }
+
+/*
+ * Make a new empty array, of the passed size, of a type that takes the
+ * priority of ap1 and ap2 into account.
+ *
+ * If `out` is non-NULL, memory overlap is checked with ap1 and ap2, and an
+ * updateifcopy temporary array may be returned. If `result` is non-NULL, the
+ * output array to be returned (`out` if non-NULL and the newly allocated array
```

Comment:
I must admit I do not understand what is the point of possibly passing on `result` to store the output in when the output is also returned. Since this PR is just about moving stuff so it can be used in multiple places, I think it is better to leave it as is here, but if you have the energy for a clean-up follow-up PR, that might be an idea...

Reply:
`result` is the output ndarray to be returned - either `out` or a freshly allocated ndarray (let's call it `retval`). `retval` is the ndarray to be iterated over. If writeback semantics are active then `retval` is not `out` rather it is a new ndarray for iterating, and `retval->base == out`. Then toward the end of the function the writeback semantics are resolved, `retval` is discared, and the function returns `return`.

Reply:
Thanks for the explanation! If in a future PR you get to touch this file again, maybe add it to the comments...